cdap-app-fabric/src/main/java/io/cdap/cdap/internal/tether/TetherServerHandler.java

```diff
@@ -0,0 +1,247 @@
+      return;
+    }
+
+    List<TetherControlMessage> commands = new ArrayList<>();
+    MessageFetcher fetcher = messagingContext.getMessageFetcher();
+    TopicId topic = new TopicId(NamespaceId.SYSTEM.getNamespace(), TETHERING_TOPIC_PREFIX + peer);
+    String lastMessageId = null;
+    try (CloseableIterator<Message> iterator =
+           fetcher.fetch(topic.getNamespace(), topic.getTopic(), 1, lastMessageIds.get(topic.getTopic()))) {
+      while (iterator.hasNext()) {
+        Message message = iterator.next();
+        TetherControlMessage controlMessage = GSON.fromJson(message.getPayloadAsString(StandardCharsets.UTF_8),
+                                                            TetherControlMessage.class);
+        commands.add(controlMessage);
+        lastMessageId = message.getId();
+      }
+    } catch (TopicNotFoundException e) {
+      LOG.warn("Received control connection from peer {} that's not tethered", peer);
+    }
+
+    if (lastMessageId != null) {
+      // Update the last message id for the topic if we read any messages
+      lastMessageIds.put(topic.getTopic(), lastMessageId);
+      String finalLastMessageId = lastMessageId;
+      TransactionRunners.run(transactionRunner, context -> {
+        AppMetadataStore.create(context).persistSubscriberState(topic.getTopic(), SUBSCRIBER, finalLastMessageId);
+      });
+    }
+
+    if (commands.isEmpty()) {
+      commands.add(new TetherControlMessage(TetherControlMessage.Type.KEEPALIVE, null));
+    }
+    Type type = new TypeToken<List<TetherControlMessage>>() { }.getType();
+    responder.sendJson(HttpResponseStatus.OK, GSON.toJson(commands, type));
+  }
+
+  /**
+   * Creates a tether with a client.
+   */
+  @POST
+  @Path("/tethering/connect")
+  public void createTether(FullHttpRequest request, HttpResponder responder)
+    throws NotImplementedException, IOException {
+    checkTetherServerEnabled();
+
+    String content = request.content().toString(StandardCharsets.UTF_8);
+    TetherConnectionRequest tetherRequest = GSON.fromJson(content, TetherConnectionRequest.class);
+    TopicId topicId = new TopicId(NamespaceId.SYSTEM.getNamespace(), "tethering_" + tetherRequest.getInstance());
+    try {
+      messagingService.createTopic(new TopicMetadata(topicId, Collections.emptyMap()));
+    } catch (TopicAlreadyExistsException e) {
+      LOG.warn("Topic {} already exists", topicId);
+    } catch (IOException e) {
+      LOG.error("Failed to create topic {}", topicId, e);
+      throw e;
+    }
+
+    // We don't need to keep track of the client metadata on the server side.
+    PeerMetadata peerMetadata = new PeerMetadata(tetherRequest.getNamespaces(), Collections.emptyMap());
+    // We don't store the peer endpoint on the server side because the connection is initiated by the client.
+    PeerInfo peerInfo = new PeerInfo(tetherRequest.getInstance(), null, TetherStatus.PENDING, peerMetadata);
+    store.addPeer(peerInfo);
```

Comment:
could this throw any exceptions? If so, may be we should catch it and delete the topic created in finally?

Reply:
Fixed to delete the topic when addPeer() throws an exception